### ISSUE TITLE
--nofilesystem= wont hide the folder completely

### ISSUE THREAD

**Issue by danieru98** [flatpak/flatpak#2013]
## Linux distribution and version
Lubuntu 16.04
## Flatpak version
Flatpak 0.99.3
## Description of the problem
If you override a flatpak application to have access only to one folder in your home directory, for example with:
`--nofilesystem=home`
`--filesystem=~/Goxel`
But then decide you actually don't want that flatpak application to have access to that folder anymore and you do: `--nofilesystem=~/Goxel` to try to remove access to this folder. Even after restarting the application, it will still be able to see the folder "~/Goxel" but wont be able to see anything inside it or write on it as expected.
So the problem is once you allow access to a folder that was previously completely hidden to a flatpak application, you can not from the command line completely hide it again. because even if you do use `--nofilesystem=` the name will still be visible.

My suggestion is to either have another option to "clear" an override, or have `--nofilesystem=` also hide the name of the directory, at least for non XDG directories.
## Steps to reproduce
Any application has this issue but using Goxel as an example:

> flatpak install --user flathub io.github.guillaumechereau.Goxel
> flatpak --user override --nofilesystem=home io.github.guillaumechereau.Goxel
> flatpak --user override --filesystem=~/Goxel io.github.guillaumechereau.Goxel

After this ~/.local/share/flatpak/overrides/io.github.guillaumechereau.Goxel will look like this:
> [Context]
> filesystems=~/Goxel;!home;

And indeed if io.github.guillaumechereau.Goxel is run, only ~/Goxel will be visible and writable.
Now if we try to remove access to ~/Goxel with:
> flatpak --user override --nofilesystem=~/Goxel io.github.guillaumechereau.Goxel

~/.local/share/flatpak/overrides/io.github.guillaumechereau.Goxel will now look like this:
> [Context]
> filesystems=!~/Goxel;!home;

And io.github.guillaumechereau.Goxel will still be able to see ~/Goxel, but not see it's actual content or write in it.

Curently the only way to completely hide ~/Goxel is go to ~/.local/share/flatpak/overrides/io.github.guillaumechereau.Goxel and manually edit this file to remove `!~/Goxel;` so it looks like this:
> [Context]
> filesystems=!home;

But there should be a way to clear the override without having to manually edit configuration files or make `nofilesystems=` hide the folder completely.

**Comment by danieru98** [comment]
A few updates: I upgraded all the packages in my Lubuntu 16.04 and updated Flatpak to 1.0.0
Tested this bug again the same way and it still happens. I was looking at the GTK3 file picker, but tested also with --command=sh and i could see the folder there too, but not the contents of course.

**Comment by alexlarsson** [comment]
Yeah, this is a bit tricky. In the case of --filesystem=/foo --nofilesystem=/foo/bar (i.e. you want access to something, but not some subdirectory of that) we implement this by bind-mounting an empty thing on top of /foo/bar.  This is what happens in you case.

I guess we need to be more smart about detecting the case were you couldn't see /foo/bar anyway, so there is no need to hide it.

**Comment by danieru98** [comment]
> I guess we need to be more smart about detecting the case were you couldn't see /foo/bar anyway, so there is no need to hide it.
if you do only --nofilesystem=/foo then /foo/bar is not visible, but if you also do --nofilesystem=/foo/bar then /foo/bar becomes visible (only the basic attributes like name owner permissions, inside will appear empty)

**Comment by vinipsmaker** [comment]
> But there should be a way to clear the override without having to manually edit configuration files

I was bitten by this issue as well. I decided to gave access to a folder, as in:

```bash
flatpak override --user --filesystem=xdg-download/foobar app
```

But then changed my mind and tried to remove that directory (using `--nofilesystem`), to only then see that the overrides for that app has been permanently dirtied.

I'd like an option to clear an override. I don't think `--nofilesystem` should clear any override, as that would have security implications, but an extra option should be provided.

**Comment by refi64** [comment]
There's always --reset if that's the only override.

**Comment by vinipsmaker** [comment]
> There's always --reset if that's the only override.

It's not the only override. Manually editing `~/.local/share/flatpak/overrides` is a more general workaround.
